File: folder_size.cs
// folder_size.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class FolderSize
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static string HumanReadable(long size)
    {
        string[] units = { "B", "KB", "MB", "GB", "TB" };
        double s = size;
        int i = 0;
        while (s >= 1024 && i < units.Length - 1) { s /= 1024; i++; }
        return $"{s:F1} {units[i]}";
    }

    static void GetFolderSize(string dir, bool recursive, int depth, int maxDepth,
                              bool excludeHidden, bool verbose,
                              Dictionary<string, long> sizeMap)
    {
        long total = 0;
        try
        {
            foreach (var entry in Directory.GetFileSystemEntries(dir))
            {
                string name = Path.GetFileName(entry);
                if (excludeHidden && name.StartsWith(".")) continue;
                if (File.Exists(entry))
                {
                    var info = new FileInfo(entry);
                    total += info.Length;
                    if (verbose) Console.WriteLine($"  {entry}: {HumanReadable(info.Length)}");
                }
                else if (Directory.Exists(entry) && recursive && (maxDepth == 0 || depth < maxDepth))
                {
                    var subMap = new Dictionary<string, long>();
                    GetFolderSize(entry, recursive, depth+1, maxDepth, excludeHidden, verbose, subMap);
                    long subTotal = 0;
                    foreach (var v in subMap.Values) subTotal += v;
                    total += subTotal;
                    foreach (var kv in subMap) sizeMap[kv.Key] = kv.Value;
                }
            }
        }
        catch (UnauthorizedAccessException)
        {
            if (verbose) Console.WriteLine(Colorize($"Permission denied: {dir}", "red"));
        }
        sizeMap[dir] = total;
    }

    static void Main(string[] args)
    {
        string path = ".";
        bool recursive = true, human = false, sortFlag = false, excludeHidden = false, verbose = false;
        int top = 0, maxDepth = 0;

        for (int i = 0; i < args.Length; i++)
        {
            string arg = args[i];
            if (arg == "-p" && i+1 < args.Length) path = args[++i];
            else if (arg == "-r") recursive = true;
            else if (arg == "-h") human = true;
            else if (arg == "-s") sortFlag = true;
            else if (arg == "-t" && i+1 < args.Length) top = int.Parse(args[++i]);
            else if (arg == "-d" && i+1 < args.Length) maxDepth = int.Parse(args[++i]);
            else if (arg == "--exclude-hidden") excludeHidden = true;
            else if (arg == "-v") verbose = true;
            else if (arg == "--help") { Console.WriteLine("Help..."); return; }
            else if (path == ".") path = arg;
        }

        if (!Directory.Exists(path))
        {
            Console.WriteLine(Colorize($"Error: directory '{path}' not found", "red"));
            return;
        }

        var sizeMap = new Dictionary<string, long>();
        GetFolderSize(path, recursive, 0, maxDepth, excludeHidden, verbose, sizeMap);

        var items = sizeMap.ToList();
        if (sortFlag) items.Sort((a, b) => b.Value.CompareTo(a.Value));
        if (top > 0 && top < items.Count) items = items.Take(top).ToList();

        foreach (var kv in items)
        {
            string sizeStr = human ? HumanReadable(kv.Value) : $"{kv.Value} B";
            string color = kv.Value > 1024*1024*1024 ? "red" : kv.Value > 1024*1024 ? "yellow" : "green";
            Console.WriteLine($"{Colorize(sizeStr.PadLeft(12), color)}  {kv.Key}");
        }
    }
}

File: folder_size.java
// folder_size.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class folder_size {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static String humanReadable(long size) {
        String[] units = {"B", "KB", "MB", "GB", "TB"};
        double s = size;
        int i = 0;
        while (s >= 1024 && i < units.length-1) { s /= 1024; i++; }
        return String.format("%.1f %s", s, units[i]);
    }

    private static void getFolderSize(Path dir, boolean recursive, int depth, int maxDepth,
                                      boolean excludeHidden, boolean verbose,
                                      Map<String, Long> sizeMap) throws IOException {
        long total = 0;
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(dir)) {
            for (Path entry : stream) {
                if (excludeHidden && entry.getFileName().toString().startsWith(".")) continue;
                if (Files.isRegularFile(entry)) {
                    long sz = Files.size(entry);
                    total += sz;
                    if (verbose) System.out.println("  " + entry + ": " + humanReadable(sz));
                } else if (Files.isDirectory(entry) && recursive && (maxDepth == 0 || depth < maxDepth)) {
                    Map<String, Long> subMap = new HashMap<>();
                    getFolderSize(entry, recursive, depth+1, maxDepth, excludeHidden, verbose, subMap);
                    long subTotal = 0;
                    for (long v : subMap.values()) subTotal += v;
                    total += subTotal;
                    sizeMap.putAll(subMap);
                }
            }
        } catch (AccessDeniedException e) {
            if (verbose) System.err.println(colorize("Permission denied: " + dir, RED));
        }
        sizeMap.put(dir.toString(), total);
    }

    public static void main(String[] args) throws IOException {
        String path = ".";
        boolean recursive = true, human = false, sortFlag = false, excludeHidden = false, verbose = false;
        int top = 0, maxDepth = 0;

        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            if (arg.equals("-p") && i+1 < args.length) path = args[++i];
            else if (arg.equals("-r")) recursive = true;
            else if (arg.equals("-h")) human = true;
            else if (arg.equals("-s")) sortFlag = true;
            else if (arg.equals("-t") && i+1 < args.length) top = Integer.parseInt(args[++i]);
            else if (arg.equals("-d") && i+1 < args.length) maxDepth = Integer.parseInt(args[++i]);
            else if (arg.equals("--exclude-hidden")) excludeHidden = true;
            else if (arg.equals("-v")) verbose = true;
            else if (arg.equals("--help")) { System.out.println("Help..."); return; }
            else if (path.equals(".")) path = arg;
        }

        Path root = Paths.get(path).toAbsolutePath();
        if (!Files.isDirectory(root)) {
            System.err.println(colorize("Error: not a directory", RED));
            System.exit(1);
        }

        Map<String, Long> sizeMap = new HashMap<>();
        getFolderSize(root, recursive, 0, maxDepth, excludeHidden, verbose, sizeMap);

        List<Map.Entry<String, Long>> items = new ArrayList<>(sizeMap.entrySet());
        if (sortFlag) items.sort((a, b) -> b.getValue().compareTo(a.getValue()));
        if (top > 0 && top < items.size()) items = items.subList(0, top);

        for (var entry : items) {
            String sizeStr = human ? humanReadable(entry.getValue()) : entry.getValue() + " B";
            String color = entry.getValue() > 1024*1024*1024 ? RED :
                           entry.getValue() > 1024*1024 ? YELLOW : GREEN;
            System.out.println(colorize(String.format("%12s", sizeStr), color) + "  " + entry.getKey());
        }
    }
}
